### PR TITLE
Jl/msgpack

### PR DESCRIPTION
**Pull request by jlouis**

### CODE REVIEW

File: src/lager_logstash_app.erl [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+%% top supervisor of the tree.
+%%
+%% @spec start(StartType, StartArgs) -> {ok, Pid} |
+%%                                      {ok, Pid, State} |
+%%                                      {error, Reason}
+%%      StartType = normal | {takeover, Node} | {failover, Node}
+%%      StartArgs = term()
+%% @end
+%%--------------------------------------------------------------------
+start(_StartType, _StartArgs) ->
```
Comment (ahf) [inline src/lager_logstash_app.erl]: `start_link`.

File: src/lager_logstash_json_formatter.erl [hunk 2]
```diff
@@ -61,4 +55,6 @@ convert({K, Atom}, Acc) when is_atom(Atom) ->
 convert(Else, Acc) -> [Else | Acc].
 
 encode(jsx, Data)   -> jsx:encode(Data);
-encode(jiffy, Data) -> jiffy:encode({Data}).
+encode(jiffy, Data) -> jiffy:encode({Data});
+encode(msgpack, Data) -> msgpack:pack(Data).
```
Comment (ahf) [inline src/lager_logstash_json_formatter.erl]: We had a problem with lists and binaries with `msgpack:pack/1` in Hydra where we ended up giving it `{pack_str, none}` as option. Might be a non-problem when the receiving side isn't an Erlang program.

File: src/lager_logstash_worker.erl [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+    {ok, State}.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+connect({tcp, Host, Port}) ->
+    connect({tcp, Host, Port, ?DEFAULT_TIMEOUT});
+connect({tcp, Host, Port, Timeout} = Conf) ->
+    Opts = [binary, {active, false}, {keepalive, true}],
```
Comment (ahf) [inline src/lager_logstash_worker.erl]: `nodelay` could make sense here?

File: src/lager_logstash_worker.erl [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    ok = file:write(Fd, Payload).
+
+%% -- Reconnect Buffering ---------------------------------------
+reconnect_buf_init() -> {initializing, {0, []}}.
+
+reconnect_buf_queue(Payload, {initializing, {N, _Msgs}}) when N > 500 ->
+    %% Buffer to big, cycle!
+    {initializing, {1, [Payload]}};
+reconnect_buf_queue(Payload, {initializing, {N, Msgs}}) ->
+    {initializing, {N+1, [Payload | Msgs]}}.
```
Comment (ahf) [inline src/lager_logstash_worker.erl]: Very nice.

File: src/lager_logstash_worker.erl (hunk 3 above)
Comment (jlouis) [inline src/lager_logstash_worker.erl]: Perhaps. I'm willing to pay 40ms for automatic buffering in packets through Nagle

File: src/lager_logstash_app.erl (hunk 1 above)
Comment (jlouis) [inline src/lager_logstash_app.erl]: o.O, it is an application callback?

**Pull request merged**